tracker/watcher.go

```diff
@@ -6,74 +6,96 @@ SPDX-License-Identifier: Apache-2.0
-	"github.com/vmware-labs/reconciler-runtime/inject"
+	"github.com/vmware-labs/reconciler-runtime/manager"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	crclient "sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-type watchFunc func(gvk schema.GroupVersionKind, controller controller.Controller) error
+type watchFunc func(gvk schema.GroupVersionKind) (context.CancelFunc, error)
 
 type watcher struct {
-	tracker Tracker
+	tracker GKAwareTracker
 	watch   watchFunc
 
-	m          sync.Mutex // protects watches and controller
-	watches    map[schema.GroupKind]struct{}
-	controller controller.Controller
+	m       sync.Mutex // protects watches
+	watches map[schema.GroupKind]context.CancelFunc
 }
 
 var (
-	_ Tracker           = (*watcher)(nil)
-	_ inject.Controller = (*watcher)(nil)
+	_ Tracker = (*watcher)(nil)
 )
 
+var nopReconciler = reconcile.Func(func(context.Context, reconcile.Request) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+})
+
 // NewWatcher returns an implementation of Tracker that lets a Reconciler register a
 // particular resource as watching a resource for a particular lease duration.
 // This watch must be refreshed periodically (e.g. by a controller resync) or
 // it will expire.
-func NewWatcher(lease time.Duration, log logr.Logger, scheme *runtime.Scheme, enqueueTracked func(by client.Object, t Tracker) handler.EventHandler) Tracker {
-	tracker := New(lease, log)
-	return NewWatchingTracker(tracker, func(gvk schema.GroupVersionKind, controller controller.Controller) error {
+func NewWatcher(sm manager.SuperManager, lease time.Duration, log logr.Logger, scheme *runtime.Scheme, enqueueTracked func(by client.Object, t Tracker) handler.EventHandler) Tracker {
+	tracker := New(lease, log).(GKAwareTracker)
+	return NewWatchingTracker(tracker, func(gvk schema.GroupVersionKind) (context.CancelFunc, error) {
 		rObj, err := scheme.New(gvk)
 		obj := rObj.(crclient.Object)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		return controller.Watch(&source.Kind{Type: obj}, enqueueTracked(obj, tracker))
+		// Create a new manager with its own cache which can be stopped when watches need to be stopped.
+		mgr, err := sm.NewManager()
+		if err != nil {
+			return nil, err
+		}
+
+		ctrl, err := builder.ControllerManagedBy(mgr).Build(nopReconciler)
+		if err != nil {
+			return nil, err
+		}
+
+		// Start the manager. This will start the above controller.
+		cancel, err := sm.AddCancelable(mgr)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := ctrl.Watch(&source.Kind{Type: obj}, enqueueTracked(obj, tracker)); err != nil {
+			cancel()
+			return nil, err
+		}
```

Comment:
Switch to a metadata only watch to limit the amount of data loaded into the informer cache

```suggestion
		// Start the manager. This will start the above controller.
		cancel, err := sm.AddCancelable(mgr)
		if err != nil {
			return nil, err
		}

		_, err = builder.ControllerManagedBy(mgr).
			Watches(&source.Kind{Type: obj}, enqueueTracked(obj, tracker), builder.OnlyMetadata).
			Build(nopReconciler)
		if err != nil {
			cancel()
			return nil, err
		}
```

Reply:
Added, thanks.

Reply:
The current implementation has reverted to non-metadata-only watches. See the design doc for more information.